### PR TITLE
[WIP] Sequence Length Index to speed up initialisation of the data loader

### PR DESCRIPTION
**Pull request by shapovalov**
Some edge cases are not supported and 3 tests are failing, but overall this is the idea.

### CODE REVIEW

**Comment by shapovalov** [comment]
@davnov134 Please review.

File: uco3d/uco3d_dataset.py [hunk 1]
```diff
@@ -131,8 +368,12 @@ class UCO3DDataset:
     remove_empty_masks_poll_whole_table_threshold: int = 300_000
     preload_metadata: bool = False
     store_sql_engine: bool = False
-    # we set it manually in the constructor
-    # _index: pd.DataFrame = field(init=False)
+    use_sequence_lengths_index: bool = True
```
Comment (davnov134) [inline uco3d/uco3d_dataset.py:371]: Add a docstring for this switch.

File: uco3d/uco3d_dataset.py [hunk 2]
```diff
@@ -44,6 +45,242 @@
 
 
 _SET_LISTS_TABLE: str = "set_lists"
+_SEQUENCE_LENGTHS_TABLE: str = "sequence_lengths"
+
+
+class IndexProtocol(Protocol):
```
[the remark this answers is not in the document]
Comment (shapovalov) [reply uco3d/uco3d_dataset.py:51]: This is an implementation detail and not part of the API, and I probably should move these classes to a different file. In general, protocol does not affect the runtime, it is just a static-analysis-type-checking-time thing, so hard to break something even if you don’t understand it.

File: tests/test_dataloader.py [hunk 3]
```diff
@@ -33,6 +33,30 @@ def test_iterate_dataset(self):
         for i in load_idx:
             _ = dataset[i]
 
+    def test_get_frame_annotations(self):
```
[the remark this answers is not in the document]
Comment (shapovalov) [reply tests/test_dataloader.py:36]: I think currently the order will be different due to the way how we sort the index. Do you think it is a problem?

[the remark this answers is not in the document]
Comment (davnov134) [reply tests/test_dataloader.py:36]: I do not think the order is a problem. Regardless, the test I suggested is imho quite important and different ordering does not block it.
